include/soci/rowset.h

```diff
@@ -41,10 +41,6 @@ class rowset_iterator
     rowset_iterator(statement & st, T & define)
         : st_(&st), define_(&define)
     {
-        assert(0 != st_);
```

Comment:
I actually spent some time reading the code before making this change as it was indeed not completely obvious that it wasn't needed, but I did convince myself about it finally:

 1. `rowset_iterator` is only constructed by `rowset::begin()` and `end()`, with the latter using the default ctor (we could enforce this by making the non default ctor private probably).
 1. `rowset::begin()` just forwards to `rowset_impl::begin()`.
 1. `rowset_impl::begin()` uses ``rowset_impl::st_` which is known to never be `NULL`.

So IMHO there is no point in keeping this assert as it's tautologically true.

But, in any case, my reasoning is always the same: if we think we must guard against the pointer being null/empty here, let's do it -- but just not using `assert()` but by throwing an exception. Do you think it's worth adding the checks here and throwing if they fail?